cratedb_toolkit/testing/testcontainers/cratedb.py

```diff
@@ -101,3 +135,65 @@ def _connect(self):
         #       In `testcontainers-java`, there is the `HttpWaitStrategy`.
         # TODO: Provide a client instance.
         wait_for_logs(self, predicate="o.e.n.Node.*started", timeout=MAX_TRIES)
+
+
+class TestDrive:
+    """
+    Use different schemas for storing the subsystem database tables, and the
+    test/example data, so that they do not accidentally touch the default `doc`
+    schema.
+    """
+
+    EXT_SCHEMA = "testdrive-ext"
+    DATA_SCHEMA = "testdrive-data"
+
+    RESET_TABLES = [
+        f'"{EXT_SCHEMA}"."retention_policy"',
+        f'"{DATA_SCHEMA}"."raw_metrics"',
+        f'"{DATA_SCHEMA}"."sensor_readings"',
+        f'"{DATA_SCHEMA}"."testdrive"',
+        f'"{DATA_SCHEMA}"."foobar"',
+        f'"{DATA_SCHEMA}"."foobar_unique_single"',
+        f'"{DATA_SCHEMA}"."foobar_unique_composite"',
+        # cratedb_toolkit.io.{influxdb,mongodb}
+        '"testdrive"."demo"',
+    ]
```

Comment:
On the other hand, contrary to `CrateDBFixture`, this section is specific to the test suite for `cratedb_toolkit`, and is not meant to be shipped with `cratedb_toolkit.testing`.

Do you see a chance to decouple this and let it be configured in `tests/conftest.py`, maybe on behalf of just a bit more pytest .request / _conf / -fixture magic, but loosely coupled, so that there is no dependency path going from `cratedb_testing` to `tests/conftest.py`, and the configuration could be somehow elegantly inverted instead?

I am not sure if I am asking for too much here, or if you can follow my thoughts easily, or if the implementation would be too complicated. Please let me know.

Reply:
If reset tables aren’t need in testing module and only needed in conftest, I can simply move them there. But in this case shall I still try to remove explicit import of testing/testcontainers? It’s needed for CrateDBFixture. I probably don’t see the whole picture, but utils module is still explicitly imported there, so I don’t quite get why it’s different for testing module

Reply:
> If reset tables aren’t need in testing module and only needed in conftest, I can simply move them there.

This specific set of tables are only meant to be reset on behalf of the cratedb-toolkit test suite, which is not being shipped as part of the package. On the other hand, `cratedb_toolkit.testing` intends to bundle generic testing helpers/utilities/fixtures.

In this spirit, you made the right choice to put `CrateDBFixture` there, but I think the table definition / test suite configuration itself, now excellently bundled into the `TestDrive` container class, should stay in `tests/conftest.py`. However, it can't be there in isolation, because it will need to be picked up by the generic `CrateDBFixture` in some way, because this one actually orchestrates the container lifecycle. Can you figure out a way to make that happen elegantly?

> But in this case shall I still try to remove explicit import of testing/testcontainers? It’s needed for CrateDBFixture.

```
# File: cratedb_toolkit/testing/testcontainers/cratedb.py
from cratedb_toolkit.testing.testcontainers.util import KeepaliveContainer, asbool
```

```
# File: tests/conftest.py
from cratedb_toolkit.testing.testcontainers.cratedb import CrateDBFixture
```

Are you referring to one of those? I think both are fine in general. It should be free for every module to use generic utilities from `cratedb_toolkit.testing`, but not the other way round, at least import-wise.

> I probably don’t see the whole picture, but utils module is still explicitly imported there, so I don’t quite get why it’s different for testing module

Are you referring to this import?

```
# File: cratedb_toolkit/testing/testcontainers/cratedb.py
from testcontainers.core.waiting_utils import wait_container_is_ready, wait_for_logs
```

I think it is also perfectly fine. You can pull in all desired utitilities _into_ `conftest.py`, but the tricky part will be to define the test suite database connectivity configuration (`TestDrive`) _there_, and let it be picked up / consumed by the generic `CrateDBFixture` to be used properly at runtime, because it can't "import" something from `tests/conftest.py`.

Reply:
So, basically, I've made `CrateDBFixture.reset` method to take reset tables in as a parameter without hardcoding them anyhow. That means, `CrateDBFixture` can still live in the `cratedb_toolkit/testing/testcontainers/cratedb.py`, but the reset tables are defined in the `tests/contest.py` along with the fixtures that use `CrateDBFixture`. I hope this solves your concern with the loose coupling.

Reply:
Yeah this was exactly what I was dreaming up regarding separation of concerns. Thank you very much.